src/python/pants/backend/python/tasks/setup_py.py

```diff
@@ -74,6 +79,42 @@ def mixins(self):
   def _setup_command(self):
     return self.__setup_command
 
+  def run(self):
+    """???/ripped from the pex installer class
+
+    Used so we can control the command line, and raise an exception on failure.
```

Comment:
This isn't making sense. The one use of `SetupPyRunner` that passes `file_input`, passes 'setup.py' which is exactly the file the original `SetupPyRunner` class operates on.

Reply:
Ok, but if we don't override the run method like this, the pex installer class generates a command line like `<interpreter_path> '-' [ARGS...]`. This causes the following error message on 1c56404b4460d6ffe18f81f8e1988debe316e6af:
```bash
> ./pants test tests/python/pants_test/backend/python/tasks:python_native_code_testing -- -vsk test_python_create_universal
# ...
 tests/python/pants_test/backend/python/tasks/test_build_local_python_distributions.py::TestBuildLocalDistsNativeSources::test_python_create_universal_distribution <- pyprep/sources/e748adc87afce18d6c648d5b1450b8a60c05af63/pants_test/backend/python/tasks/test_build_local_python_distributions.py **** Failed to install python_dist_subdir (caused by: NonZeroExit("received exit code 1 during execution of `[u'/tmp/tmp8iN8hX_BUILD_ROOT/.pants.d/pants_backend_python_tasks_build_local_python_distributions_BuildLocalPythonDistributions/ded05fd9ba17/src.python.dist.universal_dist/033964736752/setup_requires_site/setup-requires-67d1c918dfca701f0821785aaba9deb2d65ed3b6-DefaultFingerprintStrategy.da39a3ee5e6b_1c758847ad64.pex', '-', u'egg_info', u'--tag-build=+67d1c918dfca701f0821785aaba9deb2d65ed3b6-DefaultFingerprintStrategy.da39a3ee5e6b_1c758847ad64', u'bdist_wheel', u'--dist-dir', u'dist']` while trying to execute `[u'/tmp/tmp8iN8hX_BUILD_ROOT/.pants.d/pants_backend_python_tasks_build_local_python_distributions_BuildLocalPythonDistributions/ded05fd9ba17/src.python.dist.universal_dist/033964736752/setup_requires_site/setup-requires-67d1c918dfca701f0821785aaba9deb2d65ed3b6-DefaultFingerprintStrategy.da39a3ee5e6b_1c758847ad64.pex', '-', u'egg_info', u'--tag-build=+67d1c918dfca701f0821785aaba9deb2d65ed3b6-DefaultFingerprintStrategy.da39a3ee5e6b_1c758847ad64', u'bdist_wheel', u'--dist-dir', u'dist']`",)
                     ):
                     stdout:
                     
                     stderr:
                     Could not open - in the environment [/tmp/tmp8iN8hX_BUILD_ROOT/.pants.d/pants_backend_python_tasks_build_local_python_distributions_BuildLocalPythonDistributions/ded05fd9ba17/src.python.dist.universal_dist/033964736752/setup_requires_site/setup-requires-67d1c918dfca701f0821785aaba9deb2d65ed3b6-DefaultFingerprintStrategy.da39a3ee5e6b_1c758847ad64.pex]: [Errno 2] No such file or directory: '-'
                     
                     
                     FAILED
# ...
```

Reply:
I left a fixme in `setup_py.py` describing what I'm stuck on in that commit -- if there's an upstream fix we should be thinking about, I can do that too.

Reply:
OK - how about not using `SetupPyRunner`. What it does:
1. ensures `wheel` and `setuptools` are present
2. executes `setup.py` [args]

You already do 1 on your own, ensure `setup.py` exists and formulate args. So, just run your pex directly against `setup.py`:
```patch
diff --git a/src/python/pants/backend/python/tasks/build_local_python_distributions.py b/src/python/pants/backend/python/tasks/build_local_python_distributions.py
index a7ee08bc2..80f96791e 100644
--- a/src/python/pants/backend/python/tasks/build_local_python_distributions.py
+++ b/src/python/pants/backend/python/tasks/build_local_python_distributions.py
@@ -22,13 +22,13 @@ from pants.backend.python.subsystems.python_native_code import (BuildSetupRequir
                                                                 SetupPyNativeTools)
 from pants.backend.python.targets.python_requirement_library import PythonRequirementLibrary
 from pants.backend.python.tasks.pex_build_util import is_local_python_dist
-from pants.backend.python.tasks.setup_py import SetupPyRunner
 from pants.base.build_environment import get_buildroot
 from pants.base.exceptions import TargetDefinitionException, TaskError
+from pants.base.workunit import WorkUnitLabel
 from pants.build_graph.address import Address
 from pants.task.task import Task
 from pants.util.collections import assert_single_element
-from pants.util.contextutil import environment_as
+from pants.util.contextutil import pushd
 from pants.util.dirutil import safe_mkdir_for, split_basename_and_dirname
 from pants.util.memo import memoized_classproperty, memoized_property
 
@@ -254,7 +254,6 @@ class BuildLocalPythonDistributions(Task):
     self._create_dist(
       dist_target,
       dist_output_dir,
-      interpreter,
       setup_py_execution_environment,
       versioned_target_fingerprint,
       is_platform_specific)
@@ -279,51 +278,47 @@ class BuildLocalPythonDistributions(Task):
 
     dist_dir_args = ['--dist-dir', self._DIST_OUTPUT_DIR]
 
-    setup_py_command = egg_info_snapshot_tag_args + bdist_whl_args + platform_args + dist_dir_args
-    return setup_py_command
-
-  def _create_dist(self, dist_tgt, dist_target_dir, interpreter,
-                   setup_py_execution_environment, snapshot_fingerprint, is_platform_specific):
+    return (['setup.py'] +
+            egg_info_snapshot_tag_args +
+            bdist_whl_args +
+            platform_args +
+            dist_dir_args)
+
+  def _create_dist(self,
+                   dist_tgt,
+                   dist_target_dir,
+                   setup_py_execution_environment,
+                   snapshot_fingerprint,
+                   is_platform_specific):
     """Create a .whl file for the specified python_distribution target."""
     self._copy_sources(dist_tgt, dist_target_dir)
 
     setup_py_snapshot_version_argv = self._generate_snapshot_bdist_wheel_argv(
       snapshot_fingerprint, is_platform_specific)
 
-    setup_requires_interpreter = PythonInterpreter(
-      binary=setup_py_execution_environment.setup_requires_pex.path(),
-      identity=interpreter.identity,
-      extras=interpreter.extras)
-
-    setup_runner = SetupPyRunner(
-      source_dir=dist_target_dir,
-      setup_command=setup_py_snapshot_version_argv,
-      # TODO: explain why this is necessary!
-      explicit_setup_filename='setup.py',
-      interpreter=setup_requires_interpreter)
-
+    setup_requires_pex = setup_py_execution_environment.setup_requires_pex
     setup_py_env = setup_py_execution_environment.as_environment()
-    with environment_as(**setup_py_env):
-      # Build a whl using SetupPyRunner. The pex installer class's run() instance method performs
-      # the install, returns whether the install was successful, and if not, displays the stdout and
-      # stderr of the setup.py invocation directly to stderr.
-      # TODO: Would an upstream pex change make sense to allow more control of the output printed to
-      # stderr when using pex as a library (because pex prints directly to the parent process's
-      # stderr on failure)? Or is there a canonical way to just capture stderr in a with statement?
-      was_installed = setup_runner.run()
-      if not was_installed:
-        raise self.BuildLocalPythonDistributionsError(
-          "Installation of python distribution from target {target} into directory {into_dir} "
-          "failed (return value of run() was: {rc!r}).\n"
-          "The chosen interpreter was: {interpreter}.\n"
-          "The execution environment was: {env}.\n"
-          "The setup command was: {command}."
-          .format(target=dist_tgt,
-                  into_dir=dist_target_dir,
-                  rc=was_installed,
-                  interpreter=setup_requires_interpreter,
-                  env=setup_py_env,
-                  command=setup_py_snapshot_version_argv))
+
+    cmd = ' '.join(setup_requires_pex.cmdline(setup_py_snapshot_version_argv))
+    with self.context.new_workunit('setup.py', cmd=cmd, labels=[WorkUnitLabel.TOOL]) as workunit:
+      with pushd(dist_target_dir):
+        result = setup_requires_pex.run(args=setup_py_snapshot_version_argv,
+                                        env=setup_py_env,
+                                        stdout=workunit.output('stdout'),
+                                        stderr=workunit.output('stderr'))
+        if result != 0:
+          raise self.BuildLocalPythonDistributionsError(
+            "Installation of python distribution from target {target} into directory {into_dir} "
+            "failed (return value of run() was: {rc!r}).\n"
+            "The chosen interpreter was: {interpreter}.\n"
+            "The execution environment was: {env}.\n"
+            "The setup command was: {command}."
+            .format(target=dist_tgt,
+                    into_dir=dist_target_dir,
+                    rc=result,
+                    interpreter=setup_requires_pex.path(),
+                    env=setup_py_env,
+                    command=setup_py_snapshot_version_argv))
 
   def _inject_synthetic_dist_requirements(self, dist, req_lib_addr):
     """Inject a synthetic requirements library that references a local wheel.
diff --git a/src/python/pants/backend/python/tasks/setup_py.py b/src/python/pants/backend/python/tasks/setup_py.py
index 170581f4e..1f2fcf4b0 100644
--- a/src/python/pants/backend/python/tasks/setup_py.py
+++ b/src/python/pants/backend/python/tasks/setup_py.py
@@ -14,10 +14,8 @@ from builtins import map, object, str, zip
 from collections import OrderedDict, defaultdict
 
 from future.utils import PY2, PY3
-from pex.executor import Executor
 from pex.installer import InstallerBase, Packager
 from pex.interpreter import PythonInterpreter
-from pex.tracer import TRACER
 from twitter.common.collections import OrderedSet
 from twitter.common.dirutil.chroot import Chroot
 
@@ -53,10 +51,8 @@ setup(**
 class SetupPyRunner(InstallerBase):
   _EXTRAS = ('setuptools', 'wheel')
 
-  # `interpreter` is one of the kwargs often passed to this constructor.
-  def __init__(self, source_dir, setup_command, explicit_setup_filename=None, **kw):
+  def __init__(self, source_dir, setup_command, **kw):
     self.__setup_command = setup_command
-    self.explicit_setup_filename = explicit_setup_filename
     super(SetupPyRunner, self).__init__(source_dir, **kw)
 
   def mixins(self):
@@ -78,34 +74,6 @@ class SetupPyRunner(InstallerBase):
   def _setup_command(self):
     return self.__setup_command
 
-  class SetupPyError(TaskError): pass
-
-  def run(self):
-    if self._installed is not None:
-      return self._installed
-
-    with TRACER.timed('Installing {}'.format(self._install_tmp), V=2):
-      if self.explicit_setup_filename:
-        command = [self._interpreter.binary, self.explicit_setup_filename] + self._setup_command()
-      else:
-        command = [self._interpreter.binary, '-'] + self._setup_command()
-      try:
-        Executor.execute(command,
-                         env=self._interpreter.sanitized_environment(),
-                         cwd=self._source_dir,
-                         stdin_payload=self.bootstrap_script.encode('ascii'))
-        self._installed = True
-      except Executor.NonZeroExit as e:
-        self._installed = False
-        name = os.path.basename(self._source_dir)
-        raise self.SetupPyError(
-          "**** Failed to install {} (caused by: {!r}\n):\n"
-          "stdout:\n{}\nstderr:\n{}\n"
-          .format(name, e, e.stdout, e.stderr))
-
-    self._postprocess()
-    return self._installed
-
 
 class TargetAncestorIterator(object):
   """Supports iteration of target ancestor lineages."""
diff --git a/src/python/pants/binaries/executable_pex_tool.py b/src/python/pants/binaries/executable_pex_tool.py
index 659cd3ece..fc1a2830c 100644
--- a/src/python/pants/binaries/executable_pex_tool.py
+++ b/src/python/pants/binaries/executable_pex_tool.py
@@ -43,7 +43,7 @@ class ExecutablePexTool(Subsystem):
     if is_executable(pex_file_path):
       return PEX(pex_file_path, interpreter)
     else:
-      pex_info = PexInfo.default()
+      pex_info = PexInfo.default(interpreter=interpreter)
       if self.entry_point is not None:
         pex_info.entry_point = self.entry_point
 
```

You then get this: ![image](https://user-images.githubusercontent.com/157586/44486006-a5222a00-a60f-11e8-9209-5078491c45f4.png)

Reply:
Upstream fix https://github.com/pantsbuild/pex/pull/543

NB though - using SetupPyRunner here was still a bit off since:
1. The current code redundantly add `wheel` and `setuptools` requirements
2. The code was reaching across to anthother task's helper without extracting it and a dedicated test for it.